### PR TITLE
Automatically sort save files, save states, and screenshots by content dir parent folder name

### PR DESCRIPTION
**Pull request by Nargash**
## Description

I've added support to allow automatically sorting save files, save states, and screenshots into folders named after the running content parent directory folder. This is very similar to the existing flags to sort saves and states by core name. The idea behind this is it will allow people to automatically sort these files by platform, rather than just a core name, without having to resort to complicated core or content directory-specific override files. I imagine many people already sort their content by platform name, and ROM sets often are packaged like that anyhow.

The changes to sort saves and states both function with core sorting as well. If a user has both enabled, they'll be stored in 
`"<base folder>/<content dir name>/<core name>"`

For example, if I am running content stored in /SNES/, with this feature we'd automatically save the sram to :/saves/SNES/SAVENAME.srm

The new config settings are:
sort_savefiles_by_content_enable
sort_savestates_by_content_enable
sort_screenshots_by_content_enable

I added menu options for sorting save files and save states. I did NOT add a menu option for sorting screenshots; this was because it does not seem to fit in the current Saving config folder and I didn't want to put it in there without feedback.

image

## Related Issues

This pull requests address issue #10436. I also tried to partially address #10666 by adding an information logging message when sort_savestates_by_content_enable or sort_savefiles_by_content_enable is enabled, since these existing settings override both sorting by cores (existing behavior) and the new sorting by content dir name.

## Related Pull Requests

N/A

## Reviewers
I'd really like feedback on how the new menu options are named. They seem a bit wordy, but I also had to differentiate them from the existing "Sort save states in folders" and "Sort Saves into Folders" options. This is also my first time working on C code and github, so apologies for any coding style issues.
[If possible @mention all the people that should review your pull request]

### CODE REVIEW

**Comment by twinaphex** [comment]
Hi @jdgleaver , would you like to review this?

**Comment by jdgleaver** [comment]
@Nargash Thanks for this! It is a great feature, and much appreciated :)

There are, however, a few issues with the PR. Apart from minor style nits/efficiency hiccups:

- You didn't try compiling the code after rebasing with master. There are build errors now!

- You didn't increase the size of the `bool_entries` array in `menu_setting.c`, which means the new settings result in out of bounds array access

- You inadvertently committed your local `retroarch.cfg` file, overwriting the default template...

Also, we have to use C-style comments, otherwise the C89 build will fail.

This diff should fix all the issues: [save_sort_fix.patch.zip](https://github.com/libretro/RetroArch/files/5095783/save_sort_fix.patch.zip)

Please could you apply this, and check that everything still works as you intended? :)

After that, please could you also do an interactive rebase and squash your commits? (this is a handy guide: http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html)

As for adding the menu option for sorting screenshots - I do think we need this, and I guess the  `Saving` menu is the only place to put it... I believe placing it directly after `Save State Compression` is probably the best option (i.e right before the block of `Write * to Content Directory` overrides)

As for the 'wordy' menu option names - these are fine, don't worry :)

**Comment by Nargash** [comment]
Thanks for the feedback and quick response, @jdgleaver. I applied your patch, and added the menu option for screenshot sorting. I recompiled on my machine and verified that all the menu options work and also update/load from the config file.

I'm still working on how to squash my 6 commits. I'm running into trouble since I merged from upstream into the fork before I initially submitted this request, and now the rebase shows all of those commits in interactive mode. I might need a few days to figure out how to do this, and will post an update when that is complete!

**Comment by jdgleaver** [comment]
Ah, that's great - thanks for applying my changes, and for testing everything. I'm very much looking forward to getting this merged - we've needed this functionality for a long time (I can't believe I never thought of it before!)

As for your rebase... Hmm... Normally you wouldn't merge from upstream (ever) - instead, it's usual to work in a new branch, like this:

```
git checkout -b new_branch
<make changes...>
git add .
git commit
git checkout master
git pull upstream master
git push origin master
git checkout new_branch
git rebase master
```

...and that leaves everything nice and clean, without merge commits :)

But that doesn't help you much! I'm afraid I'm not experienced enough with git to advise on a complex rebase like this... :(

I wonder if it might be easier to grab your current work as a patch file (you can get it here: https://patch-diff.githubusercontent.com/raw/libretro/RetroArch/pull/11212.diff), and then just 'start over' by applying it to current master...?

**Comment by i30817** [comment]
Many people store content in per rom directories in order to add extra files without mess.

**Comment by Nargash** [comment]
Hi @i30817 - hmmmm, yes if you pack content in their own folders (like a set of multi-disc .bins) then these new options probably would not be of as much benefit. Using core specific overrides to sort your saves may be the best option for that type of setup?

Since I don't believe it is possible to discern the platform from a ROM, I can't think of a better way of implementing this. That said, the new options should always result in a more granular folder structure, so I don't think it would result in games overriding each other any more often than they could have before.

Also, @jdgleaver I think I have the rebase done!

**Comment by i30817** [comment]
> Since I don't believe it is possible to discern the platform from a ROM, I can't think of a better way of implementing this. That said, the new options should always result in a more granular folder structure, so I don't think it would result in games overriding each other any more often than they could have before.

Not anymore than the current method of a single dir at least. It would need a multi-guest platform emulator, like dolphin or MAME and a name named the same way on both consoles. Rom dumpers are usually good about avoiding such situations by accident or purpose (for instance, Resident evil 4 wii and gamecube).
